lib/Controller/ReactionController.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2021 Vitor Mattos <vitor@php.rio>
+ *
+ * @author Vitor Mattos <vitor@php.rio>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Controller;
+
+use OCA\Talk\Chat\CommentsManager;
+use OCA\Talk\Chat\ReactionManager;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\Comments\NotFoundException;
+use OCP\IRequest;
+
+class ReactionController extends AEnvironmentAwareController {
+	/** @var CommentsManager */
+	private $commentsManager;
+	/** @var ReactionManager */
+	private $reactionManager;
+
+	public function __construct(string $appName,
+								IRequest $request,
+								CommentsManager $commentsManager,
+								ReactionManager $reactionManager) {
+		parent::__construct($appName, $request);
+		$this->commentsManager = $commentsManager;
+		$this->reactionManager = $reactionManager;
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @RequireParticipant
+	 * @RequireReadWriteConversation
+	 * @RequireModeratorOrNoLobby
+	 *
+	 * @param int $messageId for reaction
+	 * @param string $reaction the reaction emoji
+	 * @return DataResponse
+	 */
+	public function react(int $messageId, string $reaction): DataResponse {
+		$participant = $this->getParticipant();
+		try {
+			// Verify if messageId is of room
+			$this->commentsManager->getComment($this->getRoom(), (string) $messageId);
```

Comment:
We also need to validate that the message is actually a reactionable message?
System messages should not be reactable? Neither deleted messages nor reactions themselves?

Reply:
Also as written above the method needs to be moved back to the chatManager and then this controller should depend on the ChatController, or even better you move all this to the reactionManager and just handle a the feedback of the reactionManager in the controller?